mother/src/drive.c

```diff
@@ -0,0 +1,252 @@
+};
+
+int diffdrive_odometry_update_from_velocity(struct DiffDriveOdometry *odom, float left_vel,
+					    float right_vel, const int time);
+
+struct DiffDriveTwist {
+	float linear_x;
+	float angular_z;
+};
+
+enum DiffDriveUpdateType {
+	POSITION_FEEDBACK,
+	VELOCITY_FEEDBACK
+};
+
+struct DiffDriveConfig {
+	float wheel_separation;
+	int wheels_per_side; // actually 2, but both are controlled by 1 signal
+	float wheel_radius;
+
+	float wheel_separation_multiplier;
+	float left_wheel_radius_multiplier;
+	float right_wheel_radius_multiplier;
+	int update_type;
+};
+
+struct DiffDrive {
+	const struct DiffDriveConfig config;
+	int previous_update_timestamp;
+	const struct DiffDriveOdometryConfig odometry_config;
+	struct DiffDriveOdometry *odometry;
+	int (*feedback_callback)(float *feedback_buffer, int buffer_len, int wheels_per_side);
+	int (*velocity_callback)(const float *velocity_buffer, int buffer_len, int wheels_per_side);
+};
+
+void *diffdrive_init(struct DiffDriveConfig *config)
+{
+	// TODO: Make a constructor here
+	return (void *)0xDEADBEEF;
+}
+
+void diffdrive_update(struct DiffDrive *drive, struct DiffDriveTwist command)
+{
+	// Get time since last update
+	const int now_timestamp = 0xDEADBEEF; // TODO: Change this
+	const float wheel_separation =
+		drive->config.wheel_separation_multiplier * drive->config.wheel_separation;
+	const float left_wheel_radius =
+		drive->config.left_wheel_radius_multiplier * drive->config.wheel_radius;
+	const float right_wheel_radius =
+		drive->config.right_wheel_radius_multiplier * drive->config.wheel_radius;
+
+	float linear_command = command.linear_x;
+	float angular_command = command.angular_z;
+	float left_feedback_mean = 0.0;
+	float right_feedback_mean = 0.0;
+
+	const int feedback_buffer_size = drive->config.wheels_per_side * 2;
+	// First N elements => Left
+	// Next N elements =>  Right
+	float *feedback =
+		k_malloc(sizeof(float) * feedback_buffer_size); // FIXME: Replace with kmalloc
+	if (drive->feedback_callback(feedback, feedback_buffer_size,
+				     drive->config.wheels_per_side)) {
+		// ERROR: Something went wrong while getting feedback
+	}
+	for (int i = 0; i < drive->config.wheels_per_side; i++) {
+		const float left_feedback = feedback[i];
+		const float right_feedback = feedback[drive->config.wheels_per_side + i];
+
+		if (isnan(left_feedback) || isnan(right_feedback)) {
+			// ERROR: One of the wheels gives invalid feedback
+		}
+
+		left_feedback_mean += left_feedback;
+		right_feedback_mean += right_feedback;
+	}
+	k_free(feedback);
+	left_feedback_mean /= drive->config.wheels_per_side;
+	right_feedback_mean /= drive->config.wheels_per_side;
+
+	if (drive->config.update_type | POSITION_FEEDBACK) {
+		// TODO: odometry update from position
+	} else {
+		diffdrive_odometry_update_from_velocity(
+			drive->odometry,
+			left_feedback_mean * left_wheel_radius * 1.0f,   // FIXME period.seconds()
```

Comment:
What do you need here specifically? @kknives

Reply:
The 1.0f needed to be replaced by a time period in seconds. That's the time taken by the update function last time. The ROS implementation takes that as a function argument. That's what I have done in https://github.com/waseemR02/kyvernitis/pull/21/commits/5c2b659d98cf3a351eb5196937e86c4826232308

We can choose to do something else, because the parent class takes care of what happens during the first class in ROS. Our user will have to manually handle that.